internal/controller/account_resolver.go

```diff
@@ -49,6 +52,66 @@ func (this authGwErrorResponse) String() string {
 	return b.String()
 }
 
+type ExpirableCachedAccountIdResolver struct {
+	AccountIdResolver
+	cache 		*lru.Cache
+	cacheTTL 	time.Duration
+	errorTTL 	time.Duration
+}
+
+
+func NewExpirableCachedAccountIdResolver(baseResolver AccountIdResolver, cacheSize int, cacheTTL, errorTTL time.Duration)(AccountIdResolver, error) {
+	cache, err := lru.New(cacheSize)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ExpirableCachedAccountIdResolver{
+		AccountIdResolver: 	baseResolver,
+		cache:				cache,
+		cacheTTL: 			cacheTTL,
+		errorTTL: 			errorTTL,
+	},nil
+}
+
+func (ecar *ExpirableCachedAccountIdResolver) MapClientIdToAccountId(ctx context.Context, clientID domain.ClientID) (domain.Identity, domain.AccountID, domain.OrgID, error){
+	//Check cache 
+	cached, ok := ecar.cache.Get(clientID)
+	if ok {
+		cachedResult := cached.(cachedResult)
+		//Check if cached result is still valid 
+		if time.Since(cachedResult.timestamp) < ecar.cacheTTL {
+			return cachedResult.identity, cachedResult.accountID, cachedResult.orgID, nil
+		}
+		if time.Since(cachedResult.timestamp) < ecar.errorTTL {
+			//if cach err is whin the erro ttl return it
+			return "","","", cachedResult.err
+		}
```

Comment:
I think this logic needs to be tweaked a bit.

I think you need to check the timestamp against the `cacheTTL`  if the cached error field is nil.  Likewise, I think you need to check the timestamp against the `errorTTL` if the cached error field is not nil.

The way it is now, I think this will return a `nil` error even if a even error response is cached since `cacheTTL` is likely to be larger than `errorTTL`.

I hope that makes sense.